src/routes/files.ts

```diff
@@ -1,14 +1,351 @@
+
+  // Fetch existing documents to merge new document
+  const directoryKeys = Array.from(directoryModifications.keys())
+  const storeDirectoryDocuments: StoreDirectoryDocument[] = []
+  if (directoryKeys.length) {
+    const directoryFetchResult = await dataStore.fetch({ keys: directoryKeys })
+
+    // Prepare directory documents
+    directoryFetchResult.rows.forEach(row => {
+      const directoryKey = row.key
+      const directoryModification = directoryModifications.get(directoryKey)
+
+      if (directoryModification) {
+        if ('doc' in row) {
+          let existingDirectory: StoreDirectoryDocument
+
+          try {
+            existingDirectory = asStoreDirectoryDocument(row.doc)
+          } catch (err) {
+            const apiClientError: ApiClientError = {
+              status: 422,
+              message:
+                `Unable to write files under '${directoryKey}'. ` +
+                `Existing document is not a directory.`
+            }
+            throw apiClientError
+          }
+
+          const directoryDocument: StoreDirectoryDocument = mergeDirectoryFilePointers(
+            existingDirectory,
+            directoryModification
+          )
+
+          // Update document
+          storeDirectoryDocuments.push(directoryDocument)
+        } else {
+          // Insert document
+          storeDirectoryDocuments.push({
+            ...directoryModification,
+            _id: directoryKey
+          })
+        }
+      }
+    })
+  }
+
+  // Prepare Repos Documents:
+
+  const repoKeys = Array.from(repoModifications.keys())
+  const repoFetchResult = await dataStore.fetch({ keys: repoKeys })
+  const storeRepoDocuments: StoreRootDocument[] = []
+
+  // Prepare repo (root) documents
+  repoFetchResult.rows.forEach(row => {
+    const repoKey = row.key
+    const repoModification = repoModifications.get(repoKey)
+
+    if (repoModification) {
+      if ('doc' in row) {
+        let existingRepo: StoreRootDocument
+
+        try {
+          existingRepo = asStoreRootDocument(row.doc)
+        } catch (err) {
+          const apiClientError: ApiClientError = {
+            status: 422,
+            message:
+              `Unable to write files under '${repoKey}'. ` +
+              `Document is not a repo.`
+          }
+          throw apiClientError
+        }
+
+        const repoDocument: StoreRootDocument = mergeDirectoryFilePointers(
+          existingRepo,
+          repoModification
+        )
+
+        storeRepoDocuments.push(repoDocument)
+      } else {
+        // Insert document
+        const newRepo: StoreRootDocument = {
+          _id: repoKey,
+          size: 0,
+          sizeLastCreated: 0,
+          maxSize: 0,
+          ...repoModification
+        }
+        storeRepoDocuments.push(newRepo)
+      }
+    }
+  })
+
+  // Write:
+
+  const results = await dataStore.bulk({
+    docs: [
+      ...storeFileDocuments,
+      ...storeDirectoryDocuments,
+      ...storeRepoDocuments
```

Comment:
Do not include the repo documents as part of the bulk update. Only do the repo if the bulk update has no errors